week3/movielens.Rmd

```diff
@@ -1,92 +1,194 @@
+  ggplot(aes(x = num_ratings)) +
+  geom_density(fill = "gray") + 
+  scale_x_log10() +
+  ylab("movie density")
+
+```
+
+```{r dist-mean-ratings-by-movie}
+# plot distribution of mean ratings by movie https://speakerdeck.com/jhofman/modeling-social-data-lecture-2-introduction-to-counting?slide=28
+# hint: try geom_histogram and geom_density
+movie_ratings %>%
+  ggplot(aes(x = avg_rating)) +
+  geom_histogram(bins = 50) +
+  ylab("no. of movies")
+
+movie_ratings %>%
+  ggplot(aes(x = avg_rating)) +
+  geom_density(fill = 'gray') +
+  ylab("movie_density")
+```
+
+```{r cdf-movie-pop}
+# rank movies by popularity and compute the cdf, or fraction of movies covered by the top-k moves https://speakerdeck.com/jhofman/modeling-social-data-lecture-2-introduction-to-counting?slide=30
+# hint: use dplyr's rank and arrange functions, and the base R sum and cumsum functions
+# store the result in a new data frame so you can use it in creating figure 2 from the paper below
+
+movie_ranks <- movie_ratings %>%
+  mutate(rank = row_number(desc(movie_ratings$num_ratings))) %>%
+  arrange(rank) %>%
+  mutate(cdf = cumsum(num_ratings/sum(num_ratings)))
+
+movie_ranks %>% head
+
+# plot the CDF of movie popularity
+movie_ranks %>%
+  ggplot(aes(x = rank, y = cdf)) +
+  geom_line() +
+  labs(x = "Movies Rank", y = "Views Density", title = "CDF of Movie Popularity")
+```
+
+
+# Per-user stats
+
+```{r aggregate-by-user}
+# aggregate ratings by user, computing mean and number of ratings
+
+user_ratings <- ratings %>% 
+  group_by(user_id) %>%
+  summarize(num_ratings = n(), avg_ratings = mean(rating))
+
+user_ratings %>% head()
+```
+
+```{r dist-user-activity}
+# plot distribution of user activity (= number of ratings the user made)
+# hint: try a log scale here
+user_ratings %>% 
+  ggplot(aes(x = num_ratings)) + 
+  geom_histogram(bins = 100) +
+  scale_x_log10() +
+  ylab("no. of users")
+
+user_ratings %>%
+  ggplot(aes(x = num_ratings)) +
+  geom_density(fill = 'gray') +
+  scale_x_log10() + 
+  ylab("user density")
+
+```
+
+# Anatomy of the long tail
+
+```{r long-tail}
+# generate the equivalent of figure 2 of this paper:
+# https://5harad.com/papers/long_tail.pdf
+
+# Specifically, for the subset of users who rated at least 10 movies,
+# produce a plot that shows the fraction of users satisfied (vertical
+# axis) as a function of inventory size (horizontal axis). We will
+# define "satisfied" as follows: an individual user is satisfied p% of
+# the time at inventory of size k if at least p% of the movies they
+# rated are contained in the top k most popular movies. As in the
+# paper, produce one curve for the 100% user satisfaction level and
+# another for 90%---do not, however, bother implementing the null
+# model (shown in the dashed lines).
+
+
+# create a table that has 90th and 100th percentile movies for each user
+user_movie <- inner_join(ratings, movie_ranks, by = 'movie_id') %>%
+  select(user_id, rank) %>%
+  group_by(user_id) %>%

```

Comment:
Would you want to include all users in this analysis? Or would you want to filter for users who don't have enough data (e.g., filter those who have not rated > 10 movies?)